pkg/controller/statusmanager/machineconfig_status.go

```diff
@@ -0,0 +1,248 @@
+			continue
+		}
+		if _, ok := newlyRenderedMachineConfigMap[role]; !ok {
+			newlyRenderedMachineConfigMap[role] = sets.Set[string]{}
+		}
+		newlyRenderedMachineConfigMap[role].Insert(mc.Name)
+	}
+	status.Lock()
+	var err error
+	if status.renderedMachineConfigs == nil {
+		status.renderedMachineConfigs, err = status.getLastRenderedMachineConfigState()
+		if err != nil {
+			status.Unlock()
+			klog.Errorf("failed to get rendered machine config state: %v", err)
+			return
+		}
+	}
+	// When new and existing rendered machine configs are same, then no changes. so return it now.
+	if reflect.DeepEqual(newlyRenderedMachineConfigMap, status.renderedMachineConfigs) {
+		status.Unlock()
+		return
+	}
+	renderedMachineConfigMap, err := status.getLastRenderedMachineConfigState()
+	if err != nil {
+		status.Unlock()
+		klog.Errorf("failed to get rendered machine config state: %v", err)
+		return
+	}
+	// Find out if any newly deleted machine configs and update status.machineConfigsBeingRemoved cache.
+	machineConfigsBeingRemoved := make(map[string]sets.Set[string])
+	for role, renderedMachineConfigs := range renderedMachineConfigMap {
+		if _, ok := newlyRenderedMachineConfigMap[role]; !ok {
+			machineConfigsBeingRemoved[role] = sets.Set[string]{}
+			machineConfigsBeingRemoved[role].Insert(renderedMachineConfigs.UnsortedList()...)
+		} else {
+			deleted := renderedMachineConfigs.Difference(newlyRenderedMachineConfigMap[role])
+			machineConfigsBeingRemoved[role].Insert(deleted.UnsortedList()...)
+		}
+	}
+	if !reflect.DeepEqual(machineConfigsBeingRemoved, status.machineConfigsBeingRemoved) {
+		status.machineConfigsBeingRemoved = machineConfigsBeingRemoved
+	}
+	var annotateUpdate bool
+	// When there are new rendered machine configs, update the annotation cache.
+	for role, newlyRenderedMachineConfigs := range newlyRenderedMachineConfigMap {
+		if _, ok := renderedMachineConfigMap[role]; !ok {
+			renderedMachineConfigMap[role] = sets.Set[string]{}
+			renderedMachineConfigMap[role].Insert(newlyRenderedMachineConfigs.UnsortedList()...)
+			annotateUpdate = true
+		} else {
+			new := newlyRenderedMachineConfigs.Difference(renderedMachineConfigMap[role])
+			if len(new) == 0 {
+				continue
+			}
+			renderedMachineConfigMap[role].Insert(new.UnsortedList()...)
+			annotateUpdate = true
+		}
+	}
+	if annotateUpdate {
+		status.renderedMachineConfigs = renderedMachineConfigMap
+		if err := status.setLastRenderedMachineConfigState(renderedMachineConfigMap); err != nil {
+			klog.Errorf("failed to set rendered machine config state: %v", err)
+		}
+	}
+	status.Unlock()
+	mcPools := &mcfgv1.MachineConfigPoolList{}
+	err = status.client.ClientFor("").CRClient().List(ctx, mcPools)
+	if err != nil {
+		klog.Errorf("failed to retrieve machine config pools: %v", err)
+		return
+	}
+	status.SetFromMachineConfigPool(mcPools.Items)
+}
+
+func (status *StatusManager) SetFromMachineConfigPool(mcPools []mcfgv1.MachineConfigPool) {
```

Comment:
Since this is the main reconciliation loop, I think it is a good place to add doctext for the method giving a brief description on how it works. For example, is not easy to grasp how, annotating the MCs even if they are being removed until we actually confirm they have been removed, is key to handle restarts gracefully.

Reply:
done.